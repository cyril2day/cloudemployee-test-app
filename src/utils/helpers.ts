type ParamObject = {[name: string]: any }


export function normalize(parent: ParamObject) {
   if (parent && parent.children) {
      parent.children.forEach((child: any) => {
         if (!child.parentId)
            child.parentId = parent.id

         normalize(child)
      })
   }
}

export function flatten(items: ParamObject[]) {
   const temp: unknown[] = []

   items.forEach(item => {
      normalize(item)
      const { children, ...props } = item 

      temp.push(props)

      if (children && Array.isArray(children)) {
         temp.push(flatten(children))
      }
   })

   return temp.flat()
}
