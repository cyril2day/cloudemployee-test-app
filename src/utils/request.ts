import axios from 'axios'


const api = axios.create({ baseURL: 'https://cloudemployee-backend-test-app.vercel.app/api' })

api.interceptors.request.use(
   config => {
      return config
   }
)

api.interceptors.response.use(
   response => {
      if (response.data)
         return response
      else
         return Promise.reject(response)
   },
   error => Promise.reject(error.response)
)

export { api }
