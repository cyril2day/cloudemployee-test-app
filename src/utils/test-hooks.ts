import { MountingOptions } from "@vue/test-utils"
import { setActivePinia, createPinia } from 'pinia'

export type TestMountingOptions = MountingOptions<unknown>

export const StoreBeforeEach = () => {
   beforeEach(() => {
      setActivePinia(createPinia())
   })
}
