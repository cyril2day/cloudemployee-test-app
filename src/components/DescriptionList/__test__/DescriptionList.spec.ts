import { createTestingPinia } from "@pinia/testing"
import { mount } from "@vue/test-utils"
import { mergeWith } from "lodash"
import { TestMountingOptions } from "../../../utils/test-hooks"
import DescriptionListVue from "../DescriptionList.vue"


const pinia = createTestingPinia()

const data = [{ _id: '123abc', name: 'Plan 1', description: 'Sample description only', type: 'Plan', status: 'Active' }]

const createWrapper = (overrides?: TestMountingOptions) => {
   const defaultMountingOptions: TestMountingOptions = {
      global: {
         plugins: [pinia]
      },
      props: {
         data: data
      } 
   }

   return mount(DescriptionListVue, mergeWith(defaultMountingOptions, overrides))
}


describe('Description List Component Unit Test', () => {
   beforeEach(() => {
      vi.resetModules()
   })

   test(`has single props called 'data' that accepts an array`, () => {
      expect.assertions(4)
      const wrapper = createWrapper()
      const propName = 'data'

      expect(wrapper.props(propName)).toHaveLength(1)
      expect(wrapper.props(propName)).toBeTypeOf('object')
      expect(wrapper.props(propName)).toBeInstanceOf(Array)
      expect(wrapper.props('invalidPropName')).toBe(undefined)
   })

   test('contains list items as slots when passed a data', () => {
      expect.assertions(2)
      const localData = data.concat(
         { _id: '456def', name: 'Plan 2', description: 'This is the second', type: 'Plan', status: 'Active' }
      )
      const wrapper = createWrapper({
         props: {
            data: localData
         } 
      })

      expect(wrapper.findAll('dl').length).toBe(2)
      expect(wrapper.find('dl').element.classList.contains('list__wrapper')).toBe(true)
   })

   test('has indicator for num of child icon on each list', () => {
      expect.assertions(1)
      const wrapper = createWrapper()
      
      expect(wrapper.find('.list__status-circle span').element.classList.contains('list__status-child-count')).toBe(true)
   })
})


describe('Description List Component Snapshot Test', () => {
   test('Elements and styles renders correctly', () => {
      expect.assertions(1)
      const wrapper = createWrapper()

      expect(wrapper.element).toMatchSnapshot()
   })
})
