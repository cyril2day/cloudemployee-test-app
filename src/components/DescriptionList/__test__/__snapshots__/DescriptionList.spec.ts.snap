// Vitest Snapshot v1

exports[`Description List Component Snapshot Test > Elements and styles renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <dl
    class="list__wrapper"
    data-v-238432e0=""
  >
    <div
      class="list__items"
      data-v-238432e0=""
    >
      
      
    </div>
    <div
      class="list__status-indicator"
      data-v-238432e0=""
    >
      <div
        data-v-238432e0=""
      >
        <i
          class="cib-discover"
          data-v-238432e0=""
        />
      </div>
    </div>
    <div
      class="list__actions"
      data-v-238432e0=""
    >
      <button
        class="list__edit"
        data-v-238432e0=""
      >
        <i
          class="cil-pencil"
          data-v-238432e0=""
        />
      </button>
      <button
        class="list__delete"
        data-v-238432e0=""
      >
        <i
          class="cil-trash"
          data-v-238432e0=""
        />
      </button>
    </div>
  </dl>
  <dl
    class="list__wrapper"
    data-v-238432e0=""
  >
    <div
      class="list__items"
      data-v-238432e0=""
    >
      
      
    </div>
    <div
      class="list__status-indicator"
      data-v-238432e0=""
    >
      <div
        data-v-238432e0=""
      >
        <i
          class="cib-discover"
          data-v-238432e0=""
        />
      </div>
    </div>
    <div
      class="list__actions"
      data-v-238432e0=""
    >
      <button
        class="list__edit"
        data-v-238432e0=""
      >
        <i
          class="cil-pencil"
          data-v-238432e0=""
        />
      </button>
      <button
        class="list__delete"
        data-v-238432e0=""
      >
        <i
          class="cil-trash"
          data-v-238432e0=""
        />
      </button>
    </div>
  </dl>
  
</div>
`;
